File: leetcode-04/solutions/python3/0525.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        if not nums:
            return 0

        ans = 0
        sum = 0
        dict = {}

        for i in range(len(nums)):
            sum += 1 if nums[i] else -1
            if sum == 0:
                ans = i + 1
            elif sum in dict:
                ans = max(ans, i - dict[sum])
            else:
                dict[sum] = i

        return ans

File: leetcode-04/solutions/cpp/0525.cpp
class Solution {
 public:
  int findMaxLength(vector<int>& nums) {
    if (nums.empty()) return 0;

    int ans = 0;
    int sum = 0;
    unordered_map<int, int> map;

    for (int i = 0; i < nums.size(); i++) {
      sum += nums[i] ? 1 : -1;
      if (sum == 0) {
        ans = i + 1;
      } else if (map.count(sum)) {
        ans = max(ans, i - map[sum]);
      } else {
        map[sum] = i;
      }
    }

    return ans;
  }
};